csrc/preseg_passes/finalize_multidevice_domains.cpp

```diff
@@ -40,41 +40,65 @@ bool validateMeshes(Fusion* fusion) {
-// computed contiguity. This preserves both the sharding and any stride order.
-// 5. For non-resharding expressions, it moves the DIDx to the front of the
-// loop/allocation domain.
-// Note: Ideally, the loop domain can follow the logical domain and the
-// allocation domain can follow the stride order specified/inferred. However, we
-// currently require loop domain to be the same as allocation domain. This
-// behavior will be modified in the future with allocation and loop domain being
-// propagated independently.
-void setLoopAndAllocationDomain(TensorView* tv, bool is_resharding) {
+bool shouldParallelizeAllocationOnStream(TensorView* tv) {
+  if (tv->isFusionInput() || tv->isFusionOutput()) {
+    return false;
+  }
+  for (Expr* use_of_tv : tv->uses()) {
+    for (TensorView* output :
+         ir_utils::filterByType<TensorView>(use_of_tv->outputs())) {
+      if (haveDifferentShardings(tv, output, {ParallelType::Stream})) {
+        return false;
+      }
+    }
+  }
+  return true;
+}
+
+bool isLoopStreamParallelized(const TensorView* tv) {
+  return std::any_of(
+      tv->getLoopDomain().begin(),
+      tv->getLoopDomain().end(),
+      [](IterDomain* id) { return id->isStream(); });
+}
+
+// Splits the allocation domain of a TensorView if it is device or stream
+// parallelized Device parallelization is always propagated to the allocation
+// domain Stream parallelization is propagated to the allocation domain if it is
+// allocated inside a for loop.
+void shardAllocation(TensorView* tv) {
+  if (!isLoopStreamParallelized(tv) && !tv->hasDeviceMesh()) {
+    // This is required for tests such as `LayoutOpTest.SchedulerKernel` The
+    // tensorview has allocation domain disjoint from logical domain. This will
+    // currently cause errors when setting allocation domain.
+    return;
+  }
   LinkedHashMap<IterDomain*, std::optional<bool>> allocation_to_contiguity;
   for (const auto&& [id, contiguity] :
        zip(tv->getMaybeAllocationDomain(), tv->getContiguity())) {
     allocation_to_contiguity.pushBack(id, contiguity);
   }
 
   // Allocation domain should be a permutation of logical domain at this point.
+  auto loop_stream_device_view =
+      tv->getLoopDomain() | std::views::filter([](IterDomain* id) {
+        return id->isStream() || id->isDeviceDim();
+      });
   std::vector<Expr*> transform_exprs = DependencyCheck::getAllExprsBetween(
       {tv->getMaybeAllocationDomain().begin(),
        tv->getMaybeAllocationDomain().end()},
-      {tv->getLoopDomain().begin(), tv->getLoopDomain().end()});
+      {loop_stream_device_view.begin(), loop_stream_device_view.end()});
 
   for (auto* expr : transform_exprs) {
     auto* split = dynamic_cast<Split*>(expr);
     NVF_ERROR(
         split != nullptr,
         "Expected all transform exprs to be a split between allocation and "
         "loop domain during sharding propagation.");
+    if (split->outer()->isStream() &&
```

Comment:
Nit: I believe you can move this filter to loop_stream_device_view as well. This way, we put all the filters in one location.